### PR TITLE
feat(dispatch): Create dispatch decorator

### PR DESCRIPTION
**Pull request by e-schultz**
Initial implementation of a `@dispatch` decorator.

```ts
class TestClass {
      instanceProperty = 'test'
      @dispatch()
      myMethod(value) {
        return {
          type: 'TEST',
          payload: {
            value,
            instanceProperty: this.instanceProperty
          }
        }
      }
    }

```

Need to do some integration testing, and ensure that the context of `this` is preserved as expected if action creators are using it.

Feature was requested in #384 - needs some more testing, code cleanup, better typings, etc.

### CODE REVIEW

File: src/decorators/dispatch.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import {
+  NgRedux,
+} from '../components/ng-redux';
+
+
+export function dispatch(options?: any) {
```
Comment (jspengeman) [inline src/decorators/dispatch.ts]: Any plans to use `options` anywhere? Not being used anywhere (maybe required by TypeScript, not too familiar with the language yet).

Comment (e-schultz) [inline src/decorators/dispatch.ts]: I was playing around with an idea to be able to provide your own custom dispatcher if you wanted instead of needing to rely on the singleton instance of NgRedux. 

I'll probably just drop this for now, and explore adding it back later as it's not serving any purpose right now, and rather get the basic use case working well first.

File: src/decorators/dispatch.spec.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
+import 'reflect-metadata';
+import { NgZone } from '@angular/core';
+import { NgRedux } from '../components/ng-redux';
+import { dispatch } from './dispatch';
+
+class MockNgZone {
+  run = fn => fn()
+}
+
+fdescribe('@dispatch', () => {
```
Comment (SethDavenport) [inline src/decorators/dispatch.spec.ts]: we'll want to change this back to regular `describe` before merge.

**Comment by SethDavenport** [comment]
I like it - let's try and productize this and get it into 6.3

**Comment by jspengeman** [comment]
@SethDavenport do you guys have a release schedule available anywhere? Curious when 6.3 would be released. Looking at using this for a work related project.

**Comment by SethDavenport** [comment]
There's no formal release schedule. Basically when this and possibly also #389 are good to go we'll just push them out and call that 6.3.

**Comment by SethDavenport** [comment]
Checked out your branch and gave it a try on the example-app. Works nicely - however there is one subtlety: Epics.

I added the new decorator to the action service like this:

```typescript
@Injectable()
export class AnimalActions {
  static readonly LOAD_STARTED = 'LOAD_STARTED';
  static readonly LOAD_SUCCEEDED = 'LOAD_SUCCEEDED';
  static readonly LOAD_FAILED = 'LOAD_FAILED';

  @dispatch()
  loadAnimals(animalType) {
    return {
      type: AnimalActions.LOAD_STARTED,
      meta: { animalType },
    };
  }

  @dispatch()
  loadSucceeded(animalType, payload) {
    return {
      type: AnimalActions.LOAD_SUCCEEDED,
      meta: { animalType },
      payload,
    };
  }

  @dispatch()
  loadFailed(animalType, error) {
    return {
      type: AnimalActions.LOAD_FAILED,
      meta: { animalType },
      error,
    };
  }
}
```

Bear in mind that my epics look like this:

```
@Injectable()
export class AnimalEpics {
  constructor(
    private service: AnimalService,
    private actions: AnimalActions,
  ) {}

  public createEpic(animalType: AnimalType) {
    return createEpicMiddleware(this.createLoadAnimalEpic(animalType));
  }

  private createLoadAnimalEpic(animalType) {
    return action$ => action$
      .ofType(AnimalActions.LOAD_STARTED)
      .filter(({ meta }) => meta.animalType === animalType)
      .switchMap(a => this.service.getAll(animalType)
        .map(data => this.actions.loadSucceeded(animalType, data))
        .catch(response => of(this.actions.loadFailed(animalType, {
          status: '' + response.status,
        }))));
  }
}
```

**Comment by SethDavenport** [comment]
screen shot 2017-05-12 at 12 26 25 pm

**Comment by SethDavenport** [comment]
What ends up happening is that my `LOAD_SUCCEEDED` action gets dispatched twice. Once by `@dispatch` and once by redux-observable.

I suppose I could not add `@dispatch` to the LOAD_SUCCEEDED action creator but that seems like cognitive overhead. 

It seems like there are cases where you sometimes want to create an action and not dispatch it.

**Comment by SethDavenport** [comment]
In my case it's probably reasonable to identify two types of actions: ones that you dispatch from the UI and ones that are only ever handled by an epic, but that feels a bit weird.   I'm wondering if it migh make sense to have `@dispatch` create a new property such that doing this:

```typescript
class MyActionService {
  @dispatch() myAction: = (payload) => ({ type: "MYACTION", payload })
}
```

Would result in something like this:

```typescript
// Just creates the action, doesn't dispatch it.
const myAction = myActionService.myAction(payload);

// Function generated by the decorator.
myActionService.dispatch_myAction(payload);
```

E.g. changing the decorator to something like this:

```typescript
// ...
const dispatchDescriptor: PropertyDescriptor = {
        get: () => wrapped,
        set: (setMethod) => originalMethod = setMethod,
      }
      Object.defineProperty(target, `dispatch_${key}`, dispatchDescriptor)
// ...

**Comment by SethDavenport** [comment]
Or perhaps this needn't be the default behaviour, but instead a parameter:

```typescript
class MyActionService {
  // Calling myAction() will create and dispatch it
  @dispatch() myAction: = (payload) => ({ type: "MYACTION", payload })

  // Calling myAction() will create it
  // Calling fire_myAction() will create and dispatch it.
  @dispatch('fire') myAction: = (payload) => ({ type: "MYACTION", payload })
}
```

**Comment by SethDavenport** [comment]
@e-schultz @jspengeman thoughts?

**Comment by e-schultz** [comment]
This is something I've also run into when having my `ActionServices` handle the dispatch, even without this decorator. 

Would fall into the case of my action creators doing the dispatch - triggering the epic, and then the epic would end up doing some action-creator type logic. I'd then want to pull that out into an action creator, and would have a mix and match of:

```ts
class MyActions {
  constructor(private ngRedux) {
  }

  someAction() {
    this.ngRedux.dispatch({ type: 'stuff' })
  }

  otherAction(x, y, z) {
    return { type: 'differentStuff', payload: { x, y, z } }
  }
}
```

Even without the decorator - I've run into the same bit of cognitive burden of:

* At times creating an action in an epic, is best left done to an action creator - and at times would end up creating helper functions there  - but had a nagging thought of 'well, they should be with the other action creators'
* But then in my action services, the mix & match of 'some that dispatch, some that don't can get confusing 

One of the issues with dynamically adding a property to the class though, is to get typescript to not complain - you then need to do something like:

```ts
class MyActionService {
  // Calling myAction() will create and dispatch it
  @dispatch() myAction: = (payload) => ({ type: "MYACTION", payload })

  // Calling myAction() will create it
  // Calling fire_myAction() will create and dispatch it.
  @dispatch('fire') myAction: = (payload) => ({ type: "MYACTION", payload })
  // define a property that will be a function with a signature that matches
  // or else calling actionServiceInstance._fireMyAction(payload)
  // will throw type errors 
  fire_myAction: (payload)=>void; 
}
```

which is why I'm sort of leaning towards not having the services be responsible for the dispatch - but up-to the other components/services/middleware/epics. 

Starting to feel the same way towards `ngRedux.dispatch()` within ActionServices that I do about `redux-thunk` - seems like a good idea, until it isn't, and then starts to create confusion / complexity when you don't expect it.

**Comment by SethDavenport** [comment]
Yeah ok forgot about the TS thing.  In that case maybe it makes sense to provide `@dispatch` as-is and let people decide whether or not to use it.

**Comment by SethDavenport** [comment]
:shipit: :)

**Comment by jspengeman** [comment]
@SethDavenport I am coming from a React/Redux background but previously was doing Angular 1.x so I am not _super_ familiar with the Angular/Redux ecosystem but I don't feel to concerned about the duplicate actions being dispatched (I'll explain why).

First, the `ActionService` you have seems like it is merely your action creators, I would not want my action creators responsible for dispatching the actions. I typically leave the actual dispatching to my view and utilize my action creators to create the action and then a `mapDispatchToProps` function to map the function I am calling in my view to my dispatch function. 

I feel like this seems more idiomatic then having my service dispatch actions (although, perhaps that is a more wide spread

```javascript
class MyViewComponent {  
  @select() someState;

  @dispatch()
   doAction() {      
     return someActionCreator(...);
   }
}
```

With `mapStateToProps` and `mapDispatchToProps` you have essentially what you have above but it can be harder to read then the example code. 

My main goal with the request was to have the store slightly abstracted away so that you don't have it all exposed to you. With redux-thunk and react-redux they explicitly only expose the `getState` and `dispatch` functions. 

In summary, the code examples I had read that have been posted so far seem great to me. Looks really concise.

**Comment by smithad15** [comment]
With the issue of having some action creators handle dispatch and others not, I think there is a way to handle this that could even include the decorator. 

Often times my desire is to remove knowledge of the store as much as possible from the component layer. At first blush, having action creators that auto-dispatch seems like the right way to do it, however as previously mentioned you then introduce the alternate case of epics and the added cognitive overhead of remembering which ones dispatch and which ones don't. This could be done quite easily by including a helper dispatch function of sorts on the action creator class.

```typescript
@Injectable()
export class AnimalActions {
  static readonly LOAD_STARTED = 'LOAD_STARTED';
  static readonly LOAD_SUCCEEDED = 'LOAD_SUCCEEDED';
  static readonly LOAD_FAILED = 'LOAD_FAILED';

  @dispatch()
  performAction(action) {
    return action
  }

  loadAnimals(animalType) {
    return {
      type: AnimalActions.LOAD_STARTED,
      meta: { animalType },
    };
  }

  loadSucceeded(animalType, payload) {
    return {
      type: AnimalActions.LOAD_SUCCEEDED,
      meta: { animalType },
      payload,
    };
  }

  loadFailed(animalType, error) {
    return {
      type: AnimalActions.LOAD_FAILED,
      meta: { animalType },
      error,
    };
  }
}
```

In this way you gain the separation of concerns, the ability to compose the functions together, easier testing of your action creators since they become pure, only a single impure function that you need to manage, and the abstraction of the store away from your component layer all within one import statement in your component.

```typescript
@Component({})
export class AnimalsComponent {
  
constructor(private actions: AnimalActions) {}

showAnimals() {
  this.actions.performAction(this.actions.loadAnimals('cat'));
}
```

The downside of course is that you are having to duplicate a `performAction` call for every action creator class and muddy up your components a bit, but I think that the upsides might be worth it. You could also implement it using plain old `dispatch`, which would probably be my preference for reasons of readability and easier unit testing.

```typescript
...
  constructor(private store: NgRedux<AppState>) {}

  performAction(action) {
    this.store.dispatch(action);
  }
...
```

If you are already having redux in your components by use of the `@select` decorator, then `@dispatch` seems to be a good addition to that paradigm.

Apologies for co-opting the PR thread, but there were some good points being brought up. Thoughts?

**Comment by jspengeman** [comment]
@smithad15 I think I we are saying the same things.

From my experience, action creators only create actions not dispatch them. Your `performAction` function is a thin wrapper around `@dispatch`. I am not sure it is needed as you could simply decorate a method in your view and use your _action creators_ to _create actions_. 

```javascript
@dispatch
doAction(payload) {
  return actionCreator(payload)
} 
```

In my view I would simply call `doAction`.

**Comment by SethDavenport** [comment]
@e-schultz is this gtg or are you still working on it?
